### DISCUSSION TITLE
Can I change the servicename when the tracer is flowing?

### DISCUSSION THREAD

**Discussion by calexfeng** [apache/skywalking#12328] [General]
Thank you all. I got a requirement, I want to change the servicename dynamically according to the param.
When I got {user: A}, I want to send to service-A;
When I got {user: B}, I want to send to service-B;
How to do it?

**Comment by calexfeng** [comment]
@wu-sheng honestly, thank you. Is there some same requirement?

**Comment by wu-sheng** [comment]
No, there isn't such a mechanism.

**Reply by calexfeng** [reply]
Thank you for you prompt response. Sorry to bother you, Can I make a plugin or implement some interface to change the service name?
Or add something to the source code. Here is what I brief design
![image](https://github.com/apache/skywalking/assets/169600509/351b0ac0-cadb-48e8-ac8d-bda54359be0a)
Thanks.

**Reply by wu-sheng** [reply]
I don't think this will work. The service name is propagated into upstream services. Such as svrA calls svrB, then the name(svrA) is propagated into svrB tracing context and reported in svrB's trace segments too.

**Reply by wu-sheng** [reply]
Note, What you are writing, obviously, is not a tracing and APM system's territory. 

If you are trying to do something like that, you should not consider SkyWalking.

**Reply by calexfeng** [reply]
Thank you sincerely. I'll try some other solutions.

**Comment by Superskyyy** [comment]
@calexfeng This is really about how you dynamically change the service-name config inside the tracer plugin, if you change the service name on the fly for some requests then it will probably work according to your need.

**Reply by calexfeng** [reply]
Thank you. I suppose that I need to do at least 2 things:
1. register the service name:  **RegisterGrpc.doServiceRegister**
2. change the service when the tracer send the Trace
Based on the above, I need to modify the "apm-agent-core" module.
Is that right?
Or should I make a new plugin like these?
![image](https://github.com/apache/skywalking/assets/169600509/ba7d3296-066b-499b-a645-fbda5da92179)
